src/authorize.ts

```diff
@@ -0,0 +1,175 @@
+   * constant for experimentation and testing purposes.
+   *
+   * The tenant represents the larger customer, which may include multiple authorized users.
+   */
+  mapToTenantId?: ((req: express.Request) => Promise<string>) | ((req: express.Request) => string);
+
+  /**
+   * Map the current request to a unique userId tag that can be used to retrieve the authorized token on
+   * subsequent requests.  Usually extracts the value from the output of the authorization system, or can be a
+   * constant for experimentation and testing purposes.
+   */
+  mapToUserId: ((req: express.Request) => Promise<string>) | ((req: express.Request) => string);
+
+  /**
+   * Called after the authorization process completed.
+   *
+   * Return false to indicate that the request was responded to, and that EveryAuth should not redirect the
+   * caller to the finishedUrl automatically.
+   */
+  onComplete?: (req: express.Request, res: express.Response, everyCtx: IEveryAuthContext) => Promise<boolean>;
+}
+
+/**
+ * @ignore
+ *
+ * Tries to determine the publically-visible URL for this part of the router.
+ */
+export const getHostedBaseUrl = (options: IEveryAuthOptions, req: express.Request): string => {
+  // Supplied in options; use that.
+  if (options.hostedBaseUrl) {
+    if (typeof options.hostedBaseUrl == 'string') {
+      return options.hostedBaseUrl;
+    } else if (typeof options.hostedBaseUrl == 'function') {
+      return options.hostedBaseUrl(req);
+    }
+  }
+
+  // Discovery time!
+  //
+  // Sometimes the originalUrl includes the entire request, sometimes it doesn't!
+  if (req.originalUrl.startsWith('http')) {
+    return req.originalUrl;
+  }
+
+  // req.hostname doesn't preserve the port, unfortunately, so test the Host header to see if it's present.
+  let port = '';
+  if (req.headers['host'] && req.headers['host'].includes(':')) {
+    port = `:${req.headers['host'].split(':')[1]}`;
+  }
+
+  // Return a hopefully valid URL.
+  return `${req.protocol}://${req.hostname}${port}${req.originalUrl}`;
+};
+
+/**
+ * Authorize a user with a particular service.
+ *
+ * @param serviceId The name of the service to authorize the user against.
+ * @param options Configuration options to control the behavior of this authorization.
+ * @return An Express Router
+ */
+export const authorize = (serviceId: string, options: IEveryAuthOptions): express.Router => {
+  const router = express.Router({ mergeParams: true });
+
+  const redirect = (req: express.Request, res: express.Response, ids: { userId: string; tenantId?: string }) => {
+    // Parse the finishedUrl, supporting both a plain path and a full url.
+    let finUrl: URL | undefined = undefined;
+    let fullUrl = false;
+
+    try {
+      finUrl = new URL(options.finishedUrl);
+      fullUrl = true;
+    } catch (_) {
+      // Not a fully qualified url, probably just a path.
+    }
+    if (!finUrl) {
+      finUrl = new URL(`http://localhost${options.finishedUrl}`);
```

Comment:
This should probably use the actual base URL of the request rather than hardcode localhost.

Reply:
It's a temporary hostname used just to allow us to extract out the query parameters.